platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,277 @@
+Roster changes at the gossip level influences which peers the module will work with. Because different peers may be
+at different points in consensus time, it may be that Alice has a newer roster than this node does, or Bob may have an
+older roster. Bob still needs to be able to gossip, even if he is behind. So each node will need to be able to honor
+connections from all peers in all rosters associated with rounds that have not expired.
+
+As with all other modules using rosters, Gossip must have a deterministic understanding of which roster applies to
+which round. It will receive this information from Hashgraph in the form of round metadata.
+
+### Event Intake
+
+The Event Intake System is responsible for receiving events, validating them, and emitting them in *topological order*.
+
+#### Validation
+
+One of the core responsibilities of Event Intake is to validate the events it has received. While this document does
+not specify the validation pipeline, it will define some of the primary steps involved in validation, so as to motivate
+the purpose of this module. That is, the following description is non-normative, but important for understanding the
+context within which this module operates.
+
+Event Intake receives events from gossip, or from the Event Creator module (i.e. a "self-event"). An "event" is actually
+a wrapper around the base protobuf event type (`EventCore`) that includes the core event data along with metadata that
+is not represented in protobuf or transmitted through gossip. One field populated in the metadata will be the byte[]
+that represents the protobuf-serialized form of the `EventCore`. Since gossip read this from the wire, there is no work
+to produce this `byte[]`, and since the `byte[]` will be needed for Gossip, we need to keep it anyway. But we will also
+hash this `byte[]` to produce the event hash, and store this hash in the metadata.
+
+After hashing, Event Intake will deduplicate events. It has a hash->event map allowing it to cheaply verify whether the
+event is a duplicate. If it is, the duplicate is discarded. Otherwise, the event is checked for "syntactic" correctness.
+For example, are all required fields populated, etc. While the Gossip system has already checked to ensure the payload
+of the event (its transactions) are limited in size and count, Event Intake will also check this as a safety measure in
+the event of bugs (if this check fails here, a noisy log statement should be produced, since this should never happen).
+
+If an event is valid, then we finally check the signature. Since validation and deduplication and hashing are
+significantly less expensive than signature verification, we wait on signature verification until the other steps are
+completed. The operating principle is that we want to fail fast and limit work for further stages in the pipeline.
+
+If an event has a very old birth-round that is expired, it is dropped. If a node sends a large number of expired events,
```

Comment:
Intake discards events based on them being *ancient*, not *expired*.

Reply:
During a sync exchange, if a peer sends even one event that is ancient from the receiver's perspective (let alone expired), the receiver can immediately identify this as hostile behavior. This is because during the first phase of the sync, peers exchange information about what they consider to be ancient/expired, and an honest partner will never send an event that is ancient from the other's perspective.

Gossip currently doesn't currently screen for this type of malicious behavior, but it would be a lot simpler to do this than to use heuristic methods.